### ISSUE TITLE
How to execute a function per button-click (Issue being unrelated to LeaderLine)

### ISSUE THREAD

**Issue by gumireddyprashanth** [anseki/leader-line#56]
Hi Anseki,

I'm dynamically creating lines.
here is a fiddle. [https://jsfiddle.net/ju5a2dge/9/](url)

Every time i click on the 'click' button my HTML changes and ID's change. for now i have replicated the issue using static html in fiddle. For the first time the line draws without any problem. line is draw from 1 to 3.

When i click for second build again, the html is cleared off and new set of html is added to it. 
now to draw from 4 to 3.
But im not able to remove() the existing line because the html was disconnected/or cleared off.
and im not able to draw a new line.

i tried to use counter to check if the line is existing, and do line.obj.remove().
but it doesnt work.

Im missing something here.. 

Thank you

**Comment by anseki** [comment]
Hi @gumireddyprashanth, thank you for your comment.
Before this issue, the https://github.com/anseki/leader-line/issues/45 was solved?

**Comment by gumireddyprashanth** [comment]
> 
> 
> Hi @gumireddyprashanth, thank you for your comment.
> Before this issue, the #45 was solved?

Yes bro! it worked! sorry i din mark it solved :(

**Comment by anseki** [comment]
I see.

I tried your example. Then new line was added when I clicked the document, and many lines were added when I clicked more.
So, I couldn't understand what is problem because I don't know your plan.

Also, unfortunately, your code has bugs and problems.
A portion of those:

- The button never work, the function run when you click anywhere. If you want to bind the function to the button, you have to write `$('#build').click(function() {...})` instead of `$(document).on('click', function() {...})`.
- The second `new LeaderLine` never be executed because `counter1` is not changed (is not counted-up).
- The `counter1` is meaningless because that is set to `0` in line#3 when every calling. That is, the `counter1 == 0` is always `true`.
- The `lines1` and `lines2` are renewed when every calling. That is, those don't have existing lines.
- You should not use the `$('#test')` frequently because that takes a high cost. You have to use the `$('#test')` just only once and you should save that into variable.
- The `.html('')` and `.append('...')` are duplicated.
- You should not use the `counter1 == 0`. You should use `counter1 === 0` instead.

So, what do you want to do?
For example, add new line from `#item-2` to `#item-1` when the button is clicked first time, and clear existing lines and add new line from `#item-4` to `#item3` when the button is clicked second time, and do nothing when the button is clicked third or more time.
Is this what you want?

**Comment by gumireddyprashanth** [comment]
Hi anseki,

Sorry, for a very bad fiddle, this is an updated one -> [https://jsfiddle.net/4eL8dzfk/2/](url)

Let me explain.

We let the user select few drop-down options and then create the boxes according to the user options.

1. My Box's (HTML) they are dynamic, can be any number on each time the user clicks build button.
2. Lines between the boxes can change each time the user clicks build.

What i want to achieve.
When i click build for the first time the lines been draw perfectly without any hassle.
But, when the user clicks for the second build(basically updates the boxes and lines), i should be able to clean the lines and re-draw a new one.
if the user clicks the 3rd time .. clear the lines and re-draw them with the latest json(start and end).

`var counter1 = 0;
$(document).on('click', '#build', function() {


    $('#test').show();
    $('#test').html('');
    $('#test').append('<div id="item-1" class="item">1</div><div id="item-2" class="item">2</div><div id="item-3" class="item">3</div><div id="item-4" class="item">4</div>')
    if (counter1 == 0) {
        alert('1')
        var lines1 = [{
            start: 'item-2',
            end: 'item-1'
        }];
        counter1++;
        lines1.forEach(function(line) {
            var start = document.getElementById(line.start),
                end = document.getElementById(line.end);

            if (start !== null && end !== null) {
                line.obj = new LeaderLine(start, end, {
                    color: '#002948',
                    size: 2,
                    path: 'straight',
                    startPlug: 'disc',
                    startPlugSize: 1.2,
                    startSocket: 'left',
                    endPlug: 'arrow3',
                    endPlugSize: 1.2,
                    endSocket: 'right'
                });
            }
        });
    } else if (counter1 > 0) {
        alert('2')
        lines1.forEach(function(line) {
            line.obj.remove();
        });

        $("#test").html('  <div id="item-1" class="item">1</div><div id="item-2" class="item">2</div><div id="item-3" class="item">3</div><div id="item-4" class="item">4</div>');
        var lines2 = [{
            start: 'item-4',
            end: 'item-3'
        }];

        lines2.forEach(function(line) {
            var start = document.getElementById(line.start),
                end = document.getElementById(line.end);

            if (start !== null && end !== null) {
                line.obj = new LeaderLine(start, end, {
                    color: '#002948',
                    size: 2,
                    path: 'straight',
                    startPlug: 'disc',
                    startPlugSize: 1.2,
                    startSocket: 'left',
                    endPlug: 'arrow3',
                    endPlugSize: 1.2,
                    endSocket: 'right'
                });
            }
        });
    }
})`

Thank you

**Comment by anseki** [comment]
Your fiddle didn't work...
`TypeError: lines1 is undefined`

**Comment by anseki** [comment]
Sorry, I can't understand your app yet.
However I tried to update your example.
https://jsfiddle.net/f19vow2y/

**Comment by gumireddyprashanth** [comment]
> 
> 
> Your fiddle didn't work...
> `TypeError: lines1 is undefined`

This is my problem, when i click for the second time.
` lines1.forEach(function(line) { line.obj.remove(); });`
I'm trying to remove the existing line, and it gives me an error lines1 undefined.

**Comment by anseki** [comment]
BTW, I can't understand why the HTML has to be replaced.

**Comment by gumireddyprashanth** [comment]
> 
> 
> BTW, I can't understand why the HTML has to be replaced.

The user can remove box1 or 2 or any.. and create new boxes.

Your fiddle works perfect.
Your are removing the lines before replacing the html with an new one.
` // Clear existing lines before HTML is cleared
  lines.forEach(function(line) { line.obj.remove(); });
  // Clear HTML
  $test.show().html(HTML);

But in my case, im clearing the html and and adding a new one before removing the lines.
I have to work it out!!

You are a master in javascript :)
Happy to use your library, thanks mate.

**Comment by anseki** [comment]
I think that replacing HTML code for removing and adding boxes is not good idea. You should use DOM for that.
Also, you can remove the lines after replacing HTML.
Try this: https://jsfiddle.net/sbgc8jaz/
So, could you close this issue if it was solved?

**Comment by gumireddyprashanth** [comment]
Thanks Anseki,

It worked!

b/w what do you mean by "You should use DOM for that.".
I'm replacing the HTML in DOM right? or i might doing something wrong.

can you just share a tutorial so i can refer to it on how to replace/construct the html.

Thanks make.. ill close the topic.

**Comment by anseki** [comment]
You can remove the lines via the `remove` method regardless of whether the HTML was replaced or not.
The fiddle that I showed verifies that. This works correctly without an error.

**Comment by gumireddyprashanth** [comment]
Thanks Anseki!!

**Comment by anseki** [comment]
:smile:

**Comment by gumireddyprashanth** [comment]
Hi Anseki,

Sorry to trouble you brother..
I have add more flavors to my project. :)

I have added Panzoom feature to my project. (When i zoom or drag the boxes, the lines should still be intact.) which is working fine on first click.
But when i add new lines for new html, the lines are drawing but when i drag and zoom the lines are not moving along. 
i know i have done something wrong.. not able to figure it out, may be im initializing the pan zoom at wrong place.

fiddle [https://jsfiddle.net/0n7bkhwy/2/](url)
Thanks mate.

**Comment by anseki** [comment]
That seems different issue.
Could you make new issue?
Also, if you can, could you prepare your code?
Sorry, your code is hard to read (e.g. strange indent).
The correctly writing will help you, me and more.

**Comment by gumireddyprashanth** [comment]
Ok Anseki,

Sorry, for the code.

I'm doing a finance project(banking) and i cant use the code written there. 
so, i had to replicate the issue somehow.. so had to do this.

Thank you, ill create a new issue.
